File: ProjectEuler/problem_3.rb
def largest_prime_factor(num)
  prime_factors(num).last
end

def prime_factors(num)
  factors = []

  while num > 1
    lowest_prime = factors.last || 2

    (lowest_prime..num).each do |f|
      if (num % f == 0)
        factors << f
        num /= f
        break
      end
    end
  end

  factors
end

def run
  num_tests = gets.chomp.to_i

  num_tests.times do
    num = gets.chomp.to_i
    puts largest_prime_factor(num)
  end
end

run

File: ProjectEuler/problem_3.cpp
using namespace std;
#include <vector>
#include <string>
#include <iostream>
#include <math.h>

long long int largestPrimeFactor (long long int num) {
	long long int currentLargestFactor = 2;
	long long int factor, sqrtNum;

	while (num > 1) {
    sqrtNum = sqrt(num);

		for (factor = currentLargestFactor; factor <= sqrtNum; factor++) {
			if (num % factor == 0) {
				currentLargestFactor = factor;
				num /= factor;
				break;
      } 
		}

    if (factor > sqrtNum) {
      currentLargestFactor = num;
      num = 1;
    }
	}

	return currentLargestFactor;
}

void run () {
	string input;
	int numTests;
	long long int num;

	cin >> numTests;

	for (int i = 0; i < numTests; i++) {
		cin >> num;

		cout << largestPrimeFactor(num) << endl;
	}

	return;
}

int main() {
	run();
	
	return 0;
}

